File: Days 71 - 80/Day 72/GetMostVisitedNodeCount.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DailyCodingProblem
{
	internal class Path
	{
		public HashSet<string> Nodes { get; set; } = new HashSet<string>();
		public Dictionary<char, int> LetterCounts { get; set; } = new Dictionary<char, int>();
	}

	internal class Day72
	{
		private static int Main(string[] args)
		{
			List<(int, int)> nodes = new List<(int, int)>
			{
				(0, 1), (0, 2), (2, 3), (3, 4)
			};

			Console.WriteLine(GetMostVisitedNodeCount("ABACA", nodes));
			Console.WriteLine(GetMostVisitedNodeCount("A", new List<(int, int)> { (0, 0) }));

			Console.ReadLine();

			return 0;
		}

		private static int GetMostVisitedNodeCount(string input, List<(int, int)> edges)
		{
			Dictionary<char, int> letterCounts = new Dictionary<char, int>();
			List<string> nodes = new List<string>();

			foreach (char c in input)
			{
				if (!letterCounts.ContainsKey(c))
				{
					letterCounts[c] = 0;
				}

				letterCounts[c]++;
				nodes.Add(c + letterCounts[c].ToString());
			}

			Dictionary<string, HashSet<string>> adjacencies = new Dictionary<string, HashSet<string>>();

			foreach ((int start, int end) in edges)
			{
				if (nodes[start] != nodes[end] || (nodes[start].Length == 0 && nodes[end].Length == 0))
				{
					if (!adjacencies.ContainsKey(nodes[start]))
					{
						adjacencies[nodes[start]] = new HashSet<string>();
					}

					adjacencies[nodes[start]].Add(nodes[end]);
				}
			}

			List<Path> paths = new List<Path>();
			Path graphPath = new Path();

			foreach (var node in adjacencies)
			{
				List<Path> newPaths = GetMostVisitedNodeCountHelper(graphPath, node.Key, adjacencies);
				paths.Concat(newPaths);
			}

			int maxCount = 0;

			foreach (Path path in paths)
			{
				int pathMax = 0;

				foreach (var letterCount in path.LetterCounts)
				{
					pathMax = Math.Max(pathMax, letterCount.Value);
				}

				maxCount = Math.Max(maxCount, pathMax);
			}

			const int InfiniteLoop = int.MaxValue;

			return maxCount > 0 ? maxCount : InfiniteLoop;
		}

		private static List<Path> GetMostVisitedNodeCountHelper(Path path, string node, Dictionary<string, HashSet<string>> adjacencies)
		{
			if (path.Nodes.Contains(node))
			{
				return new List<Path> { path };
			}

			Dictionary<char, int> newLetterCounts = new Dictionary<char, int>(path.LetterCounts);
			HashSet<string> newNodes = new HashSet<string>(path.Nodes) { node };

			if (!newLetterCounts.ContainsKey(node[0]))
			{
				newLetterCounts[node[0]] = 0;
			}

			newLetterCounts[node[0]]++;

			Path newPath = new Path
			{
				Nodes = newNodes,
				LetterCounts = newLetterCounts
			};

			if (!adjacencies.ContainsKey(node))
			{
				return new List<Path> { newPath };
			}

			List<Path> paths = new List<Path>();

			foreach (string child in adjacencies[node])
			{
				List<Path> newPaths = GetMostVisitedNodeCountHelper(newPath, child, adjacencies);
				paths.Concat(newPaths);
			}

			return paths;
		}
	}
}

File: Days 071 - 080/Day 72/GetMostVisitedNodeCount.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

struct Path
{
	std::unordered_set<std::string> nodes;
	std::unordered_map<char, unsigned int> letterCounts;
};

std::vector<Path> GetMostVisitedNodeCountHelper(Path path, const std::string& node, std::unordered_map<std::string, std::unordered_set<std::string>>& adjacencies)
{
	if (path.nodes.find(node) != path.nodes.end())
	{
		return { path };
	}

	std::unordered_map<char, unsigned int> newLetterCounts = path.letterCounts;
	std::unordered_set<std::string> newNodes = path.nodes;
	newNodes.insert(node);
	
	if (newLetterCounts.find(node[0]) == newLetterCounts.end())
	{
		newLetterCounts[node[0]] = 0;
	}

	newLetterCounts[node[0]]++;

	Path newPath { newNodes, newLetterCounts };

	if (adjacencies.find(node) == adjacencies.end())
	{
		return { newPath };
	}

	std::vector<Path> paths;

	for (const auto& child : adjacencies[node])
	{
		std::vector<Path> newPaths = GetMostVisitedNodeCountHelper(newPath, child, adjacencies);
		paths.insert(paths.end(), newPaths.begin(), newPaths.end());
	}

	return paths;
}

unsigned int GetMostVisitedNodeCount(const std::string& inputString, const std::vector<std::pair<int, int>>& edges)
{
	std::unordered_map<char, unsigned int> letterCounts;
	std::vector<std::string> nodes;

	for (char c : inputString)
	{
		if (letterCounts.find(c) == letterCounts.end())
		{
			letterCounts[c] = 0;
		}

		letterCounts[c]++;

		nodes.push_back(c + std::to_string(letterCounts[c]));
	}

	std::unordered_map<std::string, std::unordered_set<std::string>> adjacencies;

	for (const auto[start, end] : edges)
	{
		if (nodes[start] != nodes[end] || (nodes[start].empty() && nodes[end].empty()))
		{
			adjacencies[nodes[start]].insert(nodes[end]);
		}
	}

	std::vector<Path> paths;
	Path graphPath;

	for (const auto& node : adjacencies)
	{
		std::vector<Path> newPaths = GetMostVisitedNodeCountHelper(graphPath, node.first, adjacencies);
		paths.insert(paths.end(), newPaths.begin(), newPaths.end());
	}

	unsigned int maxCount = 0;

	for (const Path& path : paths)
	{
		unsigned int pathMax = 0;

		for (const auto& letterCount : path.letterCounts)
		{
			pathMax = std::max(pathMax, letterCount.second);
		}

		maxCount = std::max(maxCount, pathMax);
	}

	static constexpr unsigned int InfiniteLoop = std::numeric_limits<unsigned int>::max();

	return maxCount > 0 ? maxCount : InfiniteLoop;
}

int main(int argc, char* argv[])
{
	std::cout << GetMostVisitedNodeCount("ABACA", { { 0, 1 }, { 0, 2 }, { 2, 3 }, { 3, 4 } }) << "\n";
	std::cout << GetMostVisitedNodeCount("A", { { 0, 0 } }) << "\n";

	std::cin.get();

	return 0;
}